src/core/exec-invoke.c

```diff
@@ -2548,6 +2557,96 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
+        _cleanup_free_ char *proj_id_plain = NULL;
+        r = asprintf(&proj_id_plain, "%s%d", cgroup_path, type);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to generate project id for %s: %m", target_dir);
+
+        uint32_t proj_id = PRJID_CLAMP_INTO_QUOTA_RANGE(siphash24(proj_id_plain, strlen(proj_id_plain), k.bytes));
+
+        /* Check if project quotas are supported */
+#define MAX_PROJ_ID_RETRIES 10
+
+        int retries = 0;
+        while (true) {
+                if (retries >= MAX_PROJ_ID_RETRIES)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Exec quotas: Failed to generate unique project id for %s: %m", target_dir);
+
+                struct dqblk req;
+                r = quotactl_fd(fd, QCMD_FIXED(Q_GETQUOTA, PRJQUOTA), proj_id, &req);
+                if (r == -ESRCH || ERRNO_IS_NEG_NOT_SUPPORTED(r) || ERRNO_IS_NEG_PRIVILEGE(r)) {
+                        log_warning("Not applying storage quotas: project quotas are not supported for %s", target_dir);
+                        return 0;
+                } else if (r < 0)
+                        return log_debug_errno(r, "Exec quotas: Failed to query disk quota for prjid %u: %m", proj_id);
+
+                if (!FLAGS_SET(req.dqb_valid, QIF_BLIMITS) || req.dqb_bhardlimit <= 0)
+                        break;
+
+                random_bytes(&proj_id, sizeof(proj_id));
+                proj_id = PRJID_CLAMP_INTO_QUOTA_RANGE(proj_id);
+                retries++;
+        }
+
+        /* Set projid (equivalent to chattr +P -p) */
+        r = set_proj_id(fd, proj_id);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to set project id for %s: %m", target_dir);
+
+        /* Set Quotas */
+        struct statvfs disk_st;
+        r = fstatvfs(fd, &disk_st);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to get filesystem data for %s: %m", target_dir);
+
+        uint64_t block_limit = 0;
+        uint64_t inode_limit = 0;
+
+        if (ql.quota_absolute == 0 || ql.quota_scale == 0) {
+                block_limit = 1;
+                inode_limit = 1;
+        } else if (ql.quota_absolute == UINT64_MAX) {
+                int percent = UINT32_SCALE_TO_PERCENT(ql.quota_scale);
+                block_limit = (uint64_t)((disk_st.f_blocks / 100) * percent);
+                inode_limit = (uint64_t)((disk_st.f_files / 100) * percent);
+        } else {
+                if (disk_st.f_frsize <= 0)
+                       return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Exec quotas: Failed to get absolute block size for %s: %m", target_dir);
+
+                block_limit = (uint64_t)(ql.quota_absolute / disk_st.f_frsize);
+                if (block_limit > disk_st.f_blocks)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Exec quotas: Failed to set absolute block limit for %s: %m", target_dir);
+        }
```

Comment:
hmm, so I think there's some block size confusion here. iirc quota always deals in QIF_DQBLKSIZE block size, while statvfs has a more dynamic one in f_frsize. hence the calculations need to convert to bytes first before making adjustments, and then back to the right block size before applying.